src/nnet3/nnet-simple-component.cc

```diff
@@ -812,6 +809,319 @@ Component *AffineComponent::CollapseWithPrevious(
+	UpdatableComponent::Init(learning_rate);
+	linear_params_.Resize(output_dim / num_repeats, input_dim / num_repeats);
+	bias_params_.Resize(output_dim / num_repeats);
+	num_repeats_ = num_repeats;
+	KALDI_ASSERT(output_dim > 0 && input_dim > 0 && param_stddev >= 0.0);
+	linear_params_.SetRandn(); // sets to random normally distributed noise.
+	linear_params_.Scale(param_stddev);
+	bias_params_.SetRandn();
+	bias_params_.Scale(bias_stddev);
+}
+
+void RepeatedAffineComponent::Init(BaseFloat learning_rate,
+	int32 num_repeats,
+	std::string matrix_filename) {
+	UpdatableComponent::Init(learning_rate);
+	num_repeats_ = num_repeats;
+	CuMatrix<BaseFloat> mat;
+	ReadKaldiObject(matrix_filename, &mat); // will abort on failure.
+	KALDI_ASSERT(mat.NumCols() >= 2);
+	int32 input_dim = mat.NumCols() - 1, output_dim = mat.NumRows();
+	linear_params_.Resize(output_dim, input_dim);
+	bias_params_.Resize(output_dim);
+	linear_params_.CopyFromMat(mat.Range(0, output_dim, 0, input_dim));
+	bias_params_.CopyColFromMat(mat, input_dim);
+}
+
+void RepeatedAffineComponent::InitFromConfig(ConfigLine *cfl) {
+	bool ok = true;
+	BaseFloat learning_rate = learning_rate_;
+	int32 num_repeats = num_repeats_;
+	std::string matrix_filename;
+	int32 input_dim = -1, output_dim = -1;
+	cfl->GetValue("learning-rate", &learning_rate); // optional.
+	ok = ok && cfl->GetValue("num_repeats", &num_repeats);
+	if (cfl->GetValue("matrix", &matrix_filename)) {
+		Init(learning_rate, num_repeats, matrix_filename);
+		if (cfl->GetValue("input-dim", &input_dim))
+			KALDI_ASSERT(input_dim == InputDim() &&
+				"input-dim mismatch vs. matrix.");
+		if (cfl->GetValue("output-dim", &output_dim))
+			KALDI_ASSERT(output_dim == OutputDim() &&
+				"output-dim mismatch vs. matrix.");
+	} else {
+		ok = ok && cfl->GetValue("input-dim", &input_dim);
+		ok = ok && cfl->GetValue("output-dim", &output_dim);
+		KALDI_ASSERT(input_dim % num_repeats == 0 &&
+			"num-repeats must divide input-dim");
+		KALDI_ASSERT(output_dim % num_repeats == 0 &&
+			"num-repeats must divide output-dim");
+		BaseFloat param_stddev = 1.0 / std::sqrt(input_dim),
+			bias_stddev = 1.0;
+ 		cfl->GetValue("param-stddev", &param_stddev);
+		cfl->GetValue("bias-stddev", &bias_stddev);
+		Init(learning_rate, input_dim, output_dim, num_repeats,
+			param_stddev, bias_stddev);
+	}
+	if (cfl->HasUnusedValues())
+		KALDI_ERR << "Could not process these elements in initializer: "
+			<< cfl->UnusedValues();
+	if (!ok)
+		KALDI_ERR << "Bad initializer " << cfl->WholeLine();
+}
+
+void RepeatedAffineComponent::Propagate(const ComponentPrecomputedIndexes *indexes,
+    const CuMatrixBase<BaseFloat> &in,
+    CuMatrixBase<BaseFloat> *out) const {
+
+  // No need for asserts as they'll happen within the matrix operations.
+	int32 block_rows = linear_params_.NumRows();
+	int32 block_cols = linear_params_.NumCols();
+  // copies bias_params_ to each row
+  // of *out repeatly.
+	for (int i = 0; i < num_repeats_; i++) {
```

Comment:
I think it would be more efficient to create a temporary vector of the full (repeated) dimension containing copies of the bias, and then do one CopyRowsToVec call.
You can use the constructor of CuSubMatrix that takes a pointer argument to treat this vector as a matrix, so that you can call CopyRowsToVec on it to, in one call, repeat the bias_params_.